### PR TITLE
net, setup: Refactor the discovery (and store) step

### PR DESCRIPTION
**Pull request by EdDev**
**What this PR does / why we need it**:

Refactor the network setup flow discovery (and store) step.

The discovery step is not handled in the `netpod` setup flow.
It uses the nmstate current state to compute the data that needs to be stored.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
There is potential to further cleanup the code.
With the current refactor the places that may be further improved should be more visible to identify. 

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] ~~Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required~~
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] ~~Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.~~
- [ ] ~~Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered~~

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/network/setup/netpod/discover.go [hunk 1]
```diff
@@ -134,6 +142,123 @@ func (n NetPod) storePodInterfaceData(vmiSpecIface v1.Interface, ifaceState nmst
...
+			linkRoutes = append(linkRoutes, route)
+		}
+	}
+	if len(linkRoutes) == 0 {
+		return nil, fmt.Errorf("no gateway address found in routes for %s", podIfaceName)
+	}
+	return linkRoutes, nil
+}
+
+func resolveMacAddress(macAddressFromCurrent string, macAddressFromVMISpec string) (net.HardwareAddr, error) {
```
Comment (AlonaKaplan) [inline pkg/network/setup/netpod/discover.go:282]: The method is bridge binding specific. It's name is confusing and giving the feeling it is a generic method relevant to any binding,
I would move it and all the other bridge binding specific methods to a separate file.

File: pkg/network/setup/netpod/discover.go [hunk 2]
```diff
@@ -134,6 +142,123 @@ func (n NetPod) storePodInterfaceData(vmiSpecIface v1.Interface, ifaceState nmst
...
+		if err != nil {
+			return err
+		}
+		dhcpConfig.MAC = mac
+
+		linkRoutes, err := filterIPv4RoutesByInterface(currentStatus, podIfaceName)
+		if err != nil {
+			return err
+		}
+		dhcpConfig.Gateway = net.ParseIP(linkRoutes[0].NextHopAddress)
```
Comment (AlonaKaplan) [inline pkg/network/setup/netpod/discover.go:183]: How do you know the first one is the default gateway?

File: pkg/network/setup/netpod/discover.go [hunk 3]
```diff
@@ -134,6 +142,123 @@ func (n NetPod) storePodInterfaceData(vmiSpecIface v1.Interface, ifaceState nmst
...
+func filterRoutesByNonLocalDestination(linkRoutes []nmstate.Route, addr *vishnetlink.Addr) ([]nmstate.Route, error) {
+	var otherRoutes []nmstate.Route
+	for _, route := range linkRoutes {
+		_, dstIPNet, perr := net.ParseCIDR(route.Destination)
+		if perr != nil {
+			return nil, perr
+		}
+		isDefaultRoute := route.Destination == nmstate.DefaultDestinationRoute(vishnetlink.FAMILY_V4).String()
+		localDestination := dstIPNet.Contains(addr.IP)
+		if isDefaultRoute || !localDestination {
```
Comment (AlonaKaplan) [inline pkg/network/setup/netpod/discover.go:255]: Why is it required to check `isDefaultRoute`. Can the default route have a local destination?
Also, the original code checked the source is not local while here the destination is verified. Why?
```
route.Src != nil && route.Src.Equal(nic.IP.IP)
```

File: pkg/network/setup/netpod/discover.go [hunk 4]
```diff
@@ -134,6 +142,123 @@ func (n NetPod) storePodInterfaceData(vmiSpecIface v1.Interface, ifaceState nmst
...
+
+	log.Log.V(4).Infof("The generated dhcpConfig: %s\nRoutes: %+v", dhcpConfig.String(), dhcpConfig.Routes)
+	if err := cache.WriteDHCPInterfaceCache(n.cacheCreator, strconv.Itoa(n.podPID), podIfaceName, &dhcpConfig); err != nil {
+		return fmt.Errorf("failed to save DHCP configuration: %w", err)
+	}
+
+	return nil
+}
+
+func filterRoutesToDelegateThroughDHCP(otherRoutes []nmstate.Route) ([]vishnetlink.Route, error) {
```
Comment (AlonaKaplan) [inline pkg/network/setup/netpod/discover.go:223]: Why is this filtering required. I couldn't find it in the original implementation.

File: pkg/network/setup/netpod/discover.go [hunk 5]
```diff
@@ -134,6 +142,123 @@ func (n NetPod) storePodInterfaceData(vmiSpecIface v1.Interface, ifaceState nmst
...
+		}
+		dhcpConfig.MAC = mac
+
+		linkRoutes, err := filterIPv4RoutesByInterface(currentStatus, podIfaceName)
+		if err != nil {
+			return err
+		}
+		dhcpConfig.Gateway = net.ParseIP(linkRoutes[0].NextHopAddress)
+
+		otherRoutes, err := filterRoutesByNonLocalDestination(linkRoutes, addr)
```
Comment (AlonaKaplan) [inline pkg/network/setup/netpod/discover.go:185]: I have to admit it is hard for me to understand this routing filtering logic. It is different than the original logic, can you please explain why?

File: pkg/network/setup/netpod/netpod.go [hunk 6]
```diff
@@ -122,11 +123,20 @@ func (n NetPod) Setup() error {
...
+	}
+
+	if err = postDiscoveryHook(networkNames); err != nil {
 		return err
 	}
 
-	return n.config(currentStatus)
+	if err = n.config(currentStatus); err != nil {
+		log.Log.Reason(err).Errorf("failed to configure pod network")
+		return neterrors.CreateCriticalNetworkError(err)
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/network/setup/netpod/netpod.go:137]: Couldn't find a unit test for this new Critical error logic (although I prefer this to be reverted and ther ConfigState to return the Critical error).

**Comment by EdDev** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/194521edcece68ab030bd3dd522f1112fe76088d..0cac9276d014661b7b94bf7aa171e8d8bff9e870): Answered review comments from @AlonaKaplan

File: pkg/network/setup/netpod/netpod.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/network/setup/netpod/netpod.go:136]: ?

File: pkg/network/setup/configstate_test.go [hunk 7]
```diff
@@ -182,44 +148,30 @@ var _ = Describe("config state", func() {
 			})
 
 			It("runs with no current state (cache is empty)", func() {
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/network/setup/configstate_test.go:145]: Who promises it? You may have new hotplugged interfaces pending and hotplugged interfaces that failed setup - started, the rest can be finished.

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/0cac9276d014661b7b94bf7aa171e8d8bff9e870..83d8ceeef7d6075fa3f5863f0f73d2599c2cbf2f): Answered review.

File: pkg/network/setup/netpod/discover.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+		case vmiSpecIface.Macvtap != nil:
+		case vmiSpecIface.SRIOV != nil:
+		default:
+			return fmt.Errorf("undefined binding method: %v", vmiSpecIface)
+		}
+	}
+	return nil
+}
+
+func (n NetPod) storePodInterfaceData(vmiSpecIface v1.Interface, ifaceState nmstate.Interface) error {
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/network/setup/netpod/discover.go:109]: I don't agree, but won't insist.

File: pkg/network/setup/netpod/discover.go (hunk 2 above)
Comment (AlonaKaplan) [reply pkg/network/setup/netpod/discover.go]: ok, let's do it in a follow-up,

File: pkg/network/setup/network.go [hunk 9]
```diff
@@ -98,43 +98,23 @@ func (v VMNetworkConfigurator) getPhase1NICs(launcherPID *int, networks []v1.Net
...
+	for _, network := range networks {
+		iface := vmispec.LookupInterfaceByName(n.vmi.Spec.Domain.Devices.Interfaces, network.Name)
 		if iface == nil {
-			return nil, fmt.Errorf("no iface matching with network %s", networks[i].Name)
+			return fmt.Errorf("no iface matching with network %s", network.Name)
 		}
 
-		// Binding plugin, SR-IOV and Slirp devices are not part of the phases
-		if iface.Binding != nil || iface.SRIOV != nil || iface.Slirp != nil {
+		// Some bindings are not participation in phase 1.
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/network/setup/network.go:109]: It is not done...

[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/network/setup/network.go]: Ahh, sorry.
Now it should be fixed.

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/83d8ceeef7d6075fa3f5863f0f73d2599c2cbf2f..0471cb9e09c02da90574ec460a4670b7888892dc): Type fix that was missed.

**Pull request merged**